### PR TITLE
peft to opendelta migration (#434) + memory optimization (#320)

### PR DESCRIPTION
**Pull request by glerzing**
I closed https://github.com/CarperAI/trlx/pull/477 to reopen a new PR.

Feel free to ask questions or to make remarks. If you want the automated tests to be faster, I can reduce the number of models or configs to test. I mostly relied on unit tests for verification. A few things, like CausalILQLOutput, were added to make the methods more generic, and for easier automated tests.

### CODE REVIEW

File: trlx/trainer/accelerate_base_trainer.py [hunk 1]
```diff
@@ -307,11 +302,33 @@ def save_pretrained(self, directory: Optional[str] = None, **kwargs):
...
+                    for model in models:
+                        model.from_pretrained(input_dir)
+
+            self.accelerator.register_load_state_pre_hook(load_state_hook)
+
+            strict = False
+        else:
+            strict = True
+
+        self.accelerator.load_state(directory or self.config.train.checkpoint_dir, strict=strict, **kwargs)
```
Comment (glerzing) [inline trlx/trainer/accelerate_base_trainer.py:331]: For this function I'm not confident that everything I wrote makes sense. I have not tested it in a really distributed setting. And I'm not sure how useful it is because it just sets the internal state of accelerator but it does not set other things like self.model. And when not using peft, testing `trainer.load(checkpoint_dir)` was not working, because of some additional "base_model" prefix in the layer names.

**Comment by LouisCastricato** [comment]
Can you elaborate on what you mean by memory optimization beyond just the integration of PEFT?

**Comment by glerzing** [comment]
> Can you elaborate on what you mean by memory optimization beyond just the integration of PEFT?

If you check AutoModelForCausalLMWithHydraValueHead.__init__ and its Seq2Seq equivalent, when peft is used, we don't create a frozen head, because we can just bypass or temporarily deactivate the adapter. This comes from the issue #320.

**Review by jon-tow — CHANGES_REQUESTED** [review]
Hi, @glerzing! Amazing contribution 🙂
I've left some initial requests for changes and pointed out a few issues I encountered while launching some of our examples. We'll need to sort out a few things that break backward compatibility before merging.

File: trlx/models/modeling_ppo.py [hunk 2]
```diff
@@ -303,23 +315,30 @@ def forward(
...
         return self.base_model.generate(*args, **kwargs)
 
-    def state_dict(self, *args, **kwargs):
+    def state_dict(self, heads_only=False, *args, **kwargs):
         """
         Returns the state dictionary of the model. We add the state dictionary of the value head
         to the state dictionary of the wrapped model by prepending the key with `v_head.`.
         """
-        base_model_state_dict = self.base_model.state_dict(*args, **kwargs)
         v_head_state_dict = self.v_head.state_dict(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (glerzing) [reply trlx/models/modeling_ppo.py:329]: I haven't tried to run it with accelerate, and I have trouble reproducing it in Google Colab's because of other environment problems, but from your error log the problem comes from having `heads_only` before `*args`. I just pushed a correction. Do you still get an error ? Don't hesitate to double-check the functions save and load in accelerate_base_trainer.py.

**Comment by glerzing** [comment]
I tried to use the model [tiny-gpt2](https://huggingface.co/sshleifer/tiny-gpt2). It would be much faster than gpt2 for the tests, but unfortunately it looks like it causes some numerical instabilities in the test_backpropagation_and_disabling, which can fail depending on the random seed. I still replaced t5-small by google/t5-efficient-tiny, which works well.

File: trlx/trainer/accelerate_ppo_trainer.py [hunk 3]
```diff
@@ -67,12 +67,6 @@ def __init__(self, config: TRLConfig, **kwargs):
 
         self.store.clear_history()  # Clear the rollout store
 
-        # Setup a reference model when hydra heads are not used
-        if not hasattr(self.model, "frozen_head"):
-            self.ref_model = self.get_arch(self.config)
-            self.ref_model.to(self.accelerator.device)
-            self.ref_model.eval()
-
```
[the remark this answers is not in the document]
Comment (glerzing) [reply trlx/trainer/accelerate_ppo_trainer.py]: I'm ok with reverting that, the case where `num_layers_unfrozen <= 0` was indeed not well managed. But I guess it is simpler in practice to keep using forward_hydra for peft, even if it's not properly speaking using hydra heads. I also modified the code a bit by moving most of the peft processing from `AutoModelForCausalLMWithHydraValueHead` to `AutoModelForCausalLMWithValueHead`.

File: trlx/models/modeling_base.py [hunk 4]
```diff
@@ -119,24 +155,108 @@ def from_pretrained(  # noqa: max-complexity
...
+                    base_model = cls._auto_model_parent_class.from_pretrained(
+                        pretrained_model_name_or_path, *model_args, **from_pretrained_kwargs
+                    )
+
+                    if is_loaded_in_8bit:
+                        base_model = prepare_model_for_int8_training(
+                            base_model,
+                            **peft_int8_kwargs,
+                        )
+                    base_model = get_peft_model(base_model, peft_config)
```
[the remark this answers is not in the document]
Comment (glerzing) [reply trlx/models/modeling_base.py:215]: That was added.

**Review by maxreciprocate — APPROVED** [review]
LGTM!

**Review by jon-tow — CHANGES_REQUESTED** [review]
One final change! Per the Discord discussion on the 8-bit model forward args issue; let's add a guard check for 8-bit loading and notify to users that that it is still an experimental feature.

**Review by jon-tow — APPROVED** [review]
@glerzing Huge contribution and an amazing job! Thank you 🚀
